docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+## Goals
+
+Safety is a top priority. The compiler should not import APIs that would be less safe in Swift than they are in C++ and interop should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators).
+
+C++ interoperability aims to provide a clear, well-defined mapping for APIs that can be imported, and to draw a clear boundary between these APIs and APIs that cannot be imported. A user should be able to read the C++ interoperability documentation and have a good idea of both how much of their C++ API will be able to be imported in Swift and what that API will look like once it is imported. 
+
+It is a goal not to have hidden performance traps. C++ APIs should have similar performance in Swift as C++ and should be called directly when possible. Imported APIs, especially containers, should not be implicitly converted to Swift types (in contrast to the automatic Objective-C bridging for NSArray and NSString). 
+
+It is a goal to preserve the semantics (including performance semantics) that an API was written with by understanding the API’s semantics and mapping it to a Swift idiom that represents the same thing. 
+
+Swift programmers have used C++ APIs since Swift 1.0 through an Objective-C or C bridging layer. It is important that programmers are able to *incrementally* remove these (potentially huge) bridging layers and start using their C++ APIs directly. To make this transition incremental, C++ interoperability must not change the way Swift imports any existing C or Objective-C APIs, even when C++ interoperability is enabled. 
+
+It is *not* a goal to import every C++ API into Swift. APIs that are imported by Swift should feel Swifty and idiomatic. There must be an easy way to add annotations so that APIs become importable, especially when updating APIs directly is not an option (because the source code is not available). 
+
+## The approach
+
+The approach for importing C++ APIs into Swift is derived from the fact that both our goals, and C++ as a language, necessitate that not all APIs are imported, and to import an API the user must provide semantic information in some cases. Further, in order to realize our goals (especially the goals of safety and clear, well defined mappings), it is impossible to import any arbitrary C++ API. APIs that are imported by Swift must fall into one of our understood categories or be annotated with semantic information that can be used to classify them into a Swift idiom. 
```

Comment:
From this sentence it is not clear how much coverage of C++ types/APIs you expect from these categories. Not sure if you want to clarify that or not.

Reply:
> our goals, and C++ as a language, necessitate that not all APIs are imported

I don't know that this claim has been adequately supported by arguments.  I understand that importing all APIs is a hard problem and we shouldn't expect it in the near term, but nothing I've read so far would lead me to the conclusion above.

Reply:
Sure, I think I can update this. The point I'm trying to make here is that we want to _start_ from a place of not importing all APIs. In this initial version of interop that is outlined by the roadmap, not all APIs will be imported. This is necessitated by the fact that we want specific mappings for API patterns, and the fact that not all C++ APIs fit into one of these patterns. Does that make sense?

Of course, I'm not opposed to eventually supporting all C++ API patterns. If someone wanted to add support for all of these, I don't see any reason we shouldn't. The document outlines exactly how this can be done over time in the evolution section.